nodejs/lib/util/http.ts

```diff
@@ -0,0 +1,236 @@
+	static methodsWithBody = [HttpMethods.Post, HttpMethods.Put, HttpMethods.Patch];
+
+	/* Unlike for doUri, the 'rest' param here is mandatory, as it's used to generate the hosts */
+	static do(method: HttpMethods, rest: Rest, path: PathParameter, headers: Record<string, string> | null, body: unknown, params: RequestParams, callback: RequestCallback): void {
+		const uriFromHost = (typeof(path) == 'function') ? path : function(host: string) { return rest.baseUri(host) + path; };
+	
+		const currentFallback = rest._currentFallback;
+		if(currentFallback) {
+			if(currentFallback.validUntil > Date.now()) {
+				/* Use stored fallback */
+				Http.doUri(method, rest, uriFromHost(currentFallback.host), headers, body, params, (err?: ErrorInfo | null) => {
+					if(err && shouldFallback(err as ErrnoException)) {
+						/* unstore the fallback and start from the top with the default sequence */
+						rest._currentFallback = null;
+						Http.do(method, rest, path, headers, body, params, callback);
+						return;
+					}
+					callback(err);
+				});
+				return;
+			} else {
+				/* Fallback expired; remove it and fallthrough to normal sequence */
+				rest._currentFallback = null;
+			}
+		}
+	
+		const hosts = getHosts(rest);
+	
+		/* see if we have one or more than one host */
+		if(hosts.length == 1) {
+			Http.doUri(method, rest, uriFromHost(hosts[0]), headers, body, params, callback);
+			return;
+		}
+	
+		const tryAHost = (candidateHosts: Array<string>, persistOnSuccess?: boolean) => {
+			const host = candidateHosts.shift();
+			Http.doUri(method, rest, uriFromHost(host as string), headers, body, params, function(err?: ErrorInfo | null) {
+				if(err && shouldFallback(err as ErrnoException) && candidateHosts.length) {
+					tryAHost(candidateHosts, true);
+					return;
+				}
+				if(persistOnSuccess) {
+					/* RSC15f */
+					rest._currentFallback = {
+						host: host as string,
+						validUntil: Date.now() + rest.options.timeouts.fallbackRetryTimeout
+					};
+				}
+				callback(err)
+			});
+		};
+		tryAHost(hosts);
+	}
+
+	static doUri(method: HttpMethods, rest: Rest, uri: string, headers: Record<string, string> | null, body: unknown, params: RequestParams, callback: RequestCallback): void {
+		/* Will generally be making requests to one or two servers exclusively
+		* (Ably and perhaps an auth server), so for efficiency, use the
+		* foreverAgent to keep the TCP stream alive between requests where possible */
+		const agentOptions = (rest && rest.options.restAgentOptions) || Defaults.restAgentOptions;
+		const doOptions: RequestOptions = {uri, headers: headers ?? undefined, encoding: null, agentOptions: agentOptions};
+		if (body) {
+			doOptions.body = body;
+		}
+		if(params)
+			doOptions.qs = params;
+
+		doOptions.uri = uri;
+		doOptions.timeout = (rest && rest.options.timeouts || Defaults.TIMEOUTS).httpRequestTimeout;
+		request[method](doOptions, handler(uri, params, callback));
+	}
+
+	/** Http.get, Http.post, Http.put, ...
```

Comment:
This whole comment structure looks somewhat irregular... is this some kind of multi-method commentary syntax that I'm unaware of? 😕

Reply:
I think it's novel but not sure if there's a problem with it. Consumers of this class will get the typings from `http.d.ts` because this class isn't directly imported anywhere in source code, so it's only gonna be visible to people looking in this file and I think it's fairly obvious what the intended meaning is.